packages/kit/src/runtime/app/server/remote/query.js

```diff
@@ -104,3 +104,129 @@ export function query(validate_or_fn, maybe_fn) {
+ * Creates a batch query function that collects multiple calls and executes them in a single request
+ *
+ * See [Remote functions](https://svelte.dev/docs/kit/remote-functions#query.batch) for full documentation.
+ *
+ * @template Input
+ * @template Output
+ * @overload
+ * @param {'unchecked'} validate
+ * @param {(args: Input[]) => MaybePromise<Output[]>} fn
+ * @returns {RemoteQueryFunction<Input, Output>}
+ * @since 2.35
+ */
+/**
+ * Creates a batch query function that collects multiple calls and executes them in a single request
+ *
+ * See [Remote functions](https://svelte.dev/docs/kit/remote-functions#query.batch) for full documentation.
+ *
+ * @template {StandardSchemaV1} Schema
+ * @template Output
+ * @overload
+ * @param {Schema} schema
+ * @param {(args: StandardSchemaV1.InferOutput<Schema>[]) => MaybePromise<Output[]>} fn
+ * @returns {RemoteQueryFunction<StandardSchemaV1.InferInput<Schema>, Output>}
+ * @since 2.35
+ */
+/**
+ * @template Input
+ * @template Output
+ * @param {any} validate_or_fn
+ * @param {(args?: Input[]) => MaybePromise<Output[]>} [maybe_fn]
+ * @returns {RemoteQueryFunction<Input, Output>}
+ * @since 2.35
+ */
+/*@__NO_SIDE_EFFECTS__*/
+function batch(validate_or_fn, maybe_fn) {
+	/** @type {(args?: Input[]) => Output[]} */
+	const fn = maybe_fn ?? validate_or_fn;
+
+	/** @type {(arg?: any) => MaybePromise<Input>} */
+	const validate = create_validator(validate_or_fn, maybe_fn);
+
+	/** @type {RemoteInfo} */
+	const __ = { type: 'query.batch', id: '', name: '' };
+
+	/** @type {{ args: any[], resolvers: Array<{resolve: (value: any) => void, reject: (error: any) => void}>, timeoutId: any }} */
+	let batching = { args: [], resolvers: [], timeoutId: null };
+
+	/** @type {RemoteQueryFunction<Input, Output> & { __: RemoteInfo }} */
+	const wrapper = (arg) => {
+		if (prerendering) {
+			throw new Error(
+				`Cannot call query.batch '${__.name}' while prerendering, as prerendered pages need static data. Use 'prerender' from $app/server instead`
+			);
+		}
+
+		const { event, state } = get_request_store();
+
+		/** @type {Promise<any> & Partial<RemoteQuery<any>>} */
+		const promise = get_response(__.id, arg, state, () => {
+			// Collect all the calls to the same query in the same macrotask,
+			// then execute them as one backend request.
+			return new Promise((resolve, reject) => {
+				batching.args.push(arg);
+				batching.resolvers.push({ resolve, reject });
+
+				if (batching.timeoutId) {
+					clearTimeout(batching.timeoutId);
+				}
+
+				batching.timeoutId = setTimeout(async () => {
+					const batched = batching;
+					batching = { args: [], resolvers: [], timeoutId: null };
+
+					try {
+						const results = await run_remote_function(
+							event,
+							state,
+							false,
+							batched.args,
+							(array) => Promise.all(array.map(validate)),
+							fn
+						);
+						for (let i = 0; i < batched.resolvers.length; i++) {
+							batched.resolvers[i].resolve(results[i]);
+						}
```

Comment:
we should throw an error here if `results.length !== batched.args.length`. something like this?

```suggestion

						if (results.length !== batched.args.length) {
							const input = `${batched.args.length} ${batched.args.length === 1 ? 'input' : 'inputs'}`;
							const output = `${results.length} ${results.length === 1 ? 'output' : 'outputs'}`;

							throw new Error(`\`${__.name}\` was called with ${input}, but returned ${output}`);
						}

						for (let i = 0; i < batched.resolvers.length; i++) {
							batched.resolvers[i].resolve(results[i]);
						}
```

Reply:
Are you saying you don't trust our own output? 😅 For the sake of having as little client side code as possible I vote "no" or at least "dev mode only"

Reply:
This is server code. The case this handles is the one where the user's code returns an array of different length to the inputs. To take your example query...

```js
const posts = await db.sql`
  SELECT * FROM post
  WHERE slug = ANY(${slugs})
`;
```

...if you have the slugs `['foo', 'bar', 'baz']` but `bar` doesn't exist in the database, then `await getPost('bar')` will return data for `baz`, while `await getPost('baz')` will return `undefined`.

Some systems let you put an error in place of a return value, but given likely usage patterns I think it's easier to just fail all the queries